### PR TITLE
[multikueue] Manage worker cluster unavailability

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

- Keep the AdmissionCheck Active even if not all it's clusters are usable.
- Add a mechanism which keeps a workload's MultiKueue AdmissionCheckState as `Ready`  for a configured time even if the manager cannot connect to its reserving worker.

#### Which issue(s) this PR fixes:

Relates to #693

#### Does this PR introduce a user-facing change?

```release-note
MuliKueue  - Manage worker cluster unavailability
```

### CODE REVIEW

File: apis/config/v1beta1/defaults.go [hunk 1]
```diff
@@ -44,6 +44,9 @@ const (
 	DefaultQueueVisibilityUpdateIntervalSeconds int32   = 5
 	DefaultClusterQueuesMaxCount                int32   = 10
 	defaultJobFrameworkName                             = "batch/job"
+	DefaultMultiKueueGCTimeout                          = time.Minute
+	DefaultMultiKueueOrigin                             = "multikueue"
+	DefaultMultiKueueKeepReadyTimeout                   = 5 * time.Minute
```
Comment (trasc) [inline apis/config/v1beta1/defaults.go:49]: 15 min

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 2]
```diff
@@ -316,36 +362,43 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
...
 			if err != nil {
-				return err
+				return reconcile.Result{}, err
 			}
 		}
-		// drop this if we want to create new remote workloads while holding a reservation
-		return nil
+		// drop this if we want to create new remote workloads while holding a reservation.
+		// check again the connection to the remote in half the keepReadyTimeout.
+		return reconcile.Result{RequeueAfter: a.keepReadyTimeout / 2}, nil
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:377]: Does this /2 mean we do two checks before giving up? If so, maybe we should make this two a named constant, and I would suggest having 3 checks, so /3?

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 3]
```diff
@@ -279,11 +293,43 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
...
+
+	// If there is no reserving and the AC is ready, the connection with the reserving remote might
+	// be lost, keep the workload admitted for keepReadyTimeout and put it back in the queue after that.
+	if !hasReserving && acs.State == kueue.CheckStateReady {
+		if time.Now().Before(acs.LastTransitionTime.Add(a.keepReadyTimeout)) {
+			retryAfter := a.keepReadyTimeout - time.Since(acs.LastTransitionTime.Time)
+			log.V(3).Info("Reserving remote lost, retry", "retryAfter", retryAfter)
+			return reconcile.Result{RequeueAfter: retryAfter}, nil
+		} else {
+			acs.State = kueue.CheckStateRetry
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:324]: I'm not sure this is the purpose of the `CheckStateRetry`, could we just keep it Ready?

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 4]
```diff
@@ -279,11 +293,43 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
...
 		})
-		return a.client.Status().Patch(ctx, wlPatch, client.Apply, client.FieldOwner(ControllerName+"-finish"), client.ForceOwnership)
+		return reconcile.Result{}, a.client.Status().Patch(ctx, wlPatch, client.Apply, client.FieldOwner(ControllerName+"-finish"), client.ForceOwnership)
 	}
 
 	hasReserving, reservingRemote := group.FirstReserving()
 
+	// if the reserving workload is out of sync, remove it
+	if hasReserving {
+		outOfSync := group.local == nil || !equality.Semantic.DeepEqual(group.local.Spec, group.remotes[reservingRemote].Spec)
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:303]: Will this not be done by GC?

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 5]
```diff
@@ -279,11 +293,43 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
...
+			group.remotes[reservingRemote] = nil
+
+			// check if another one exists
+			hasReserving, reservingRemote = group.FirstReserving()
+		}
+	}
+
+	// If there is no reserving and the AC is ready, the connection with the reserving remote might
+	// be lost, keep the workload admitted for keepReadyTimeout and put it back in the queue after that.
+	if !hasReserving && acs.State == kueue.CheckStateReady {
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:318]: Why do we differ the logic based on !hasReserving?
I think the main use case for the timeout is to prevent requeue of a running workload (admitted locally and remotely), in case of transient connectivity loss. I thought in case of such workloads we have hasReserving=true.

**Review by mimowo — COMMENTED** [review]
Initial review, API field name suggestion, and questions to understand the flow.

File: test/integration/multikueue/multikueue_test.go [hunk 6]
```diff
@@ -560,4 +562,205 @@ var _ = ginkgo.Describe("Multikueue", func() {
...
+
+		ginkgo.By("the worker2 wl is removed since the local one no longer has a reservation", func() {
+			gomega.Eventually(func(g gomega.Gomega) {
+				createdWorkload := &kueue.Workload{}
+				g.Expect(worker2TestCluster.client.Get(worker2TestCluster.ctx, wlLookupKey, createdWorkload)).To(utiltesting.BeNotFoundError())
+			}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
+		})
+	})
+
+	ginkgo.It("Should remove the worker's workload and job when managers job is deleted", func() {
```
Comment (mimowo) [inline test/integration/multikueue/multikueue_test.go:708]: Can you explain how this test is related to the code changes? If unrelated, maybe a separate PR

File: test/integration/multikueue/multikueue_test.go [hunk 7]
```diff
@@ -560,4 +562,205 @@ var _ = ginkgo.Describe("Multikueue", func() {
...
+				g.Expect(activeCondition).To(gomega.BeComparableTo(&metav1.Condition{
+					Type:   kueuealpha.MultiKueueClusterActive,
+					Status: metav1.ConditionFalse,
+					Reason: "BadConfig",
+				}, cmpopts.IgnoreFields(metav1.Condition{}, "Message", "LastTransitionTime")))
+				disconnectedTime = activeCondition.LastTransitionTime.Time
+			}, util.Timeout, util.Interval).Should(gomega.Succeed())
+		})
+
+		ginkgo.By("waiting for the local workload admission check state to be set to pending and quotaReservatio removed", func() {
```
Comment (mimowo) [inline test/integration/multikueue/multikueue_test.go:662]: I'm wondering about using Consistency to verify the workload is not created on worker1 as long as the timeout is not reached, this would require increasing the timeout though.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 4 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:302]: Isn't this part `group.local == nil` covered by the MK GC? IIRC MK GC removes workloads on workers which don't have the local counterpart, so seems to be the case here.

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 8]
```diff
@@ -279,11 +292,43 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
...
+		outOfSync := group.local == nil || !equality.Semantic.DeepEqual(group.local.Spec, group.remotes[reservingRemote].Spec)
+		if outOfSync {
+			if err := client.IgnoreNotFound(group.RemoveRemoteObjects(ctx, reservingRemote)); err != nil {
+				log.V(2).Error(err, "Deleting out of sync remote objects", "remote", reservingRemote)
+				return reconcile.Result{}, err
+			}
+			group.remotes[reservingRemote] = nil
+
+			// check if another one exists
+			hasReserving, reservingRemote = group.FirstReserving()
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:311]: Should we not repeat the outOfSync check for the new pair: `hasReserving, reservingRemote `?

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 3 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:323]: Ok, I have two concerns with that:
1. in the [docs for Retry](https://github.com/kubernetes-sigs/kueue/blob/fa96518c293473480b0d1f45f642af23e433302e/apis/kueue/v1beta1/admissioncheck_types.go#L28C2-L30C19
) we say "A workload having at least one check in the state, will be evicted if admitted will not be considered for admission.". However, in the case of workload being temporarily unreachable it is already admitted, so I'm not sure it fits the definition, does it? 

2. In the [docs for Rejected](https://github.com/kubernetes-sigs/kueue/blob/fa96518c293473480b0d1f45f642af23e433302e/apis/kueue/v1beta1/admissioncheck_types.go#L35-L37. ) we have "NOTE: The admission behaviour is currently the same as for retry, we can consider marking the workload as "Finished" with a failure description.". Should we update the note then? I'm also concerned about the checks in code which make `Retry` and `Rejected` equivalent, like [here](https://github.com/kubernetes-sigs/kueue/blob/fa96518c293473480b0d1f45f642af23e433302e/pkg/workload/admissionchecks.go#L145), so we may need to review all the places if this still makes sense.

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 9]
```diff
@@ -279,11 +292,43 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
...
+		}
+	}
+
+	// If there is no reserving and the AC is ready, the connection with the reserving remote might
+	// be lost, keep the workload admitted for keepReadyTimeout and put it back in the queue after that.
+	if !hasReserving && acs.State == kueue.CheckStateReady {
+		remainingWaitTime := a.workerLostTimeout - time.Since(acs.LastTransitionTime.Time)
+		if remainingWaitTime > 0 {
+			log.V(3).Info("Reserving remote lost, retry", "retryAfter", remainingWaitTime)
+			return reconcile.Result{RequeueAfter: remainingWaitTime}, nil
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:321]: Requeue after 15min to check sounds bad. I think by analogy to `--retry` in curl we could have exponential backoff, or fixed delay (probably at 30s by default). Then wait for `min(remainingTime, delay)`

File: pkg/controller/admissionchecks/multikueue/workload_test.go [hunk 10]
```diff
@@ -423,6 +424,185 @@ func TestWlReconcile(t *testing.T) {
 					Obj(),
 			},
 		},
+		"the local workload is kept Ready if the keepReadyTimeout did not pass since last successful sync": {
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload_test.go:427]:
```suggestion
		"the local workload admission check Ready if the remote WorkerLostTimeout is not exceeded": {
```

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 11]
```diff
@@ -316,15 +362,19 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
...
 			if err != nil {
-				return err
+				return reconcile.Result{}, err
 			}
 		}
-		// drop this if we want to create new remote workloads while holding a reservation
-		return nil
+		// drop this if we want to create new remote workloads while holding a reservation.
+		// check again the connection to the remote in workerLostTimeout.
+		return reconcile.Result{RequeueAfter: a.workerLostTimeout}, nil
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:377]: As in the other comment, I think we should check the connection more often than in 15min.

File: pkg/controller/admissionchecks/multikueue/workload_test.go [hunk 12]
```diff
@@ -423,6 +424,185 @@ func TestWlReconcile(t *testing.T) {
...
+						Name:    "ac1",
+						State:   kueue.CheckStateReady,
+						Message: `The workload got reservation on "worker1"`,
+					}).
+					ControllerReference(batchv1.SchemeGroupVersion.WithKind("Job"), "job1", "uid1").
+					ReserveQuota(utiltesting.MakeAdmission("q1").Obj()).
+					Obj(),
+			},
+		},
+		"the local workload is set for Retry if the keepReadyTimeout did pass since last successful sync": {
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload_test.go:453]:
```suggestion
		"the local workload's admission check is set to Retry if the WorkerLostTimeout is exceeded": {
```

File: pkg/controller/admissionchecks/multikueue/workload_test.go [hunk 13]
```diff
@@ -423,6 +424,185 @@ func TestWlReconcile(t *testing.T) {
...
+						Name:    "ac1",
+						State:   kueue.CheckStateRetry,
+						Message: `Reserving remote lost`,
+					}).
+					ControllerReference(batchv1.SchemeGroupVersion.WithKind("Job"), "job1", "uid1").
+					ReserveQuota(utiltesting.MakeAdmission("q1").Obj()).
+					Obj(),
+			},
+		},
+		"the local workload is set to pendin when it is in Retry without quota": {
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload_test.go:479]:
```suggestion
		"the local workload's admission check is set to Pending when it is in Retry without quota": {
```

File: test/integration/multikueue/multikueue_test.go [hunk 14]
```diff
@@ -560,4 +562,146 @@ var _ = ginkgo.Describe("Multikueue", func() {
...
+			gomega.Expect(managerTestCluster.client.Get(managerTestCluster.ctx, wlLookupKey, managerWl)).To(gomega.Succeed())
+			gomega.Eventually(func(g gomega.Gomega) {
+				g.Expect(worker2TestCluster.client.Get(worker2TestCluster.ctx, wlLookupKey, createdWorkload)).To(gomega.Succeed())
+				g.Expect(createdWorkload.Spec).To(gomega.BeComparableTo(managerWl.Spec))
+				g.Expect(worker1TestCluster.client.Get(worker1TestCluster.ctx, wlLookupKey, createdWorkload)).To(gomega.Succeed())
+				g.Expect(createdWorkload.Spec).To(gomega.BeComparableTo(managerWl.Spec))
+			}, util.Timeout, util.Interval).Should(gomega.Succeed())
+		})
+
+		ginkgo.By("setting workload reservation in worker2, the workload is admitted in manager amd worker1 wl is removed", func() {
```
Comment (mimowo) [inline test/integration/multikueue/multikueue_test.go:613]:
```suggestion
		ginkgo.By("setting workload reservation in worker2, the workload is admitted in manager and worker1 wl is removed", func() {
```

File: test/integration/multikueue/suite_test.go [hunk 15]
```diff
@@ -46,6 +46,10 @@ import (
 	// +kubebuilder:scaffold:imports
 )
 
+const (
+	testingKeepReadyTimeout = 3 * time.Second
```
Comment (mimowo) [inline test/integration/multikueue/suite_test.go:50]:
```suggestion
	workerLostTimeout = 3 * time.Second
```

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 3 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:323]: In other words, I'm not understanding yet was is the difference between `Rejected` and `Retry`, and why `Retry` is preferred. 

After the `WorkerLostTimeout` is exceeded we should requeue the workload. Can you elaborate what is the difference between `Rejected` and `Retry`, and how the choice impacts the scenario.

Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:323]: `Rejected` is permanent, the workload is marked as `Finished` and should never go back in the queue

File: test/integration/multikueue/multikueue_test.go (hunk 7 above)
Comment (mimowo) [reply test/integration/multikueue/multikueue_test.go:662]: Before we say "let's keep it as is" - is it covered at the unit test level though? I mean, do we have some test that verifies that a workload is not requeued before the timeout is reached?

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 3 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:323]: Got it, thanks for explaining. Should we update the descriptions to indicate the difference (potentially in this PR, preparatory, or follow up)? This "A workload having at least one check in the state, will be evicted if admitted **will not be considered for admission.**" suggests to me that the workload is not retried also. Similarly, the note suggests "NOTE: The admission behaviour is currently the same as for retry".

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 9 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:321]: The 15 min will apply only if no other event triggers the reconcile, then and only then we are interested in requeuing if the reserving worker is not back.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 4 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:302]: To avoid nil pointer we can have `group.local != nil && !equality.Semantic.DeepEqual(group.local.Spec, group.remotes[reservingRemote].Spec)`, so it is not convincing argument.

> I cannot understand why we wold like to wait fro the periodic GC to run when we can do the cleanup now.

Sorry, for now I just asked the question to understand what is happening (not claiming which one is better in the comments above) because I wasn't sure of my understanding. Can you summarize the pros & cons for deleting now and leaving this to MKGC?

Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:302]: Can you?

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 8 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:311]: previously , L308 we remove the out of sync one, we check if another also has reservation.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 3 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:323]: indeed 

https://github.com/kubernetes-sigs/kueue/blob/fa96518c293473480b0d1f45f642af23e433302e/apis/kueue/v1beta1/admissioncheck_types.go#L35-L38

is outdated, it should be changed in a cleanup  PR

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 9 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:321]: Maybe I still miss some piece of information. Some questions: if we lose connectivity for 5min, do we mark the cluster a Say we have a workload which is running on worker1. Then, worker1 is marked as inactive (the AdmissionCheck is PartiallyActive). What will trigger the check to include the cluster back as active?

Also, if we don't get connectivity for 15min, do we mark such cluster as inactive?

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 8 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:311]: Reworked

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 9 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:321]: > Maybe I still miss some piece of information. Some question: if we have a workload running on worker1, and we lose connectivity for 5min, to worker1 (and worker1 is marked inactive). What will trigger the check to include the cluster back as active?

Nothing for now we have a follow-up for connection monitoring and reconnects.

> Here we are talking about workloads.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 3 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go]: #1763

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 9 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go]: Ok, I see, but there is an issue with delaying the check, that if we get unlucky once and 15min later, then re requeue. However, the cluster might have been available in-between for 14min. I don't think this is the intention. I think we should monitor if the cluster is connected, and switch its status based on that. 

If the problem is solved under " connection monitoring and reconnects.", then maybe it makes sense to combine to see the solution to temporary loss of connectivity holistically. At least update the "special notes to reviewers" with what will be covered in the coming follow up.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 4 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go]: > If left, the workload will remain queued in the worker cluster and can potentially block quota

Right, this is true in the other cases as well, I'm wondering what makes this situation different. What is the scenario leading to "group.local == nil"?

> We cannot consider a non-nil wl being Semantic.DeepEqual to a nil.

I don't think I suggested that (at least not intentionally). I'm saying (1) nil local workload -> MKGC, (2)  !Semantic.DeepEqual -> Delete remote. It is actually the current code that interprets (1) and (2) as the same (outOfSync). Unless I'm missing something.

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 16]
```diff
@@ -254,7 +257,17 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
 				log.V(2).Error(err, "Deleting remote workload", "workerCluster", rem)
 			}
 		}
-		return errors.Join(errs...)
+
+		if !workload.HasQuotaReservation(group.local) && acs.State == kueue.CheckStateRetry {
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:261]: So IIUC this code is to clear the admission check status as Pending after the workload is evicted and requeued here: https://github.com/kubernetes-sigs/kueue/blob/7ed44becf8bee6f894cb0e2a6f9cd37c98d7b575/pkg/controller/core/workload_controller.go#L237C30-L246. 

Do we currently keep the state of admission checks on eviction?

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 4 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go]: > Right, this is true in the other cases as well, I'm wondering what makes this situation different.

We can do something to avoid it. 

The function was refactored, please check it.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:261]: Only the admission check controller is managing the admission check state.

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 17]
```diff
@@ -316,15 +335,32 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
 			}
 			// update the message
 			acs.Message = fmt.Sprintf("The workload got reservation on %q", reservingRemote)
+			// update the transition time
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:338]: The comment explains perfectly readable line, but it is more important to write **why**, because it often depends on interaction with other parts of the system, which is not clear by looking at the lines below. Tbh, I don't know why it is updated.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 4 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go]: Oh, the code with the check is gone, feel free to resolve.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:261]: Hm, setting Retry to Pending feels like a job the workload_controller could do on eviction. Is it discussed in KEP?

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 9 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go]: When the watch starts, events are generated for the already exiting objects, hence at that point in time when the worker is alive for some time we wold get a reconcile also.

Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go]: But we still would not observe the cluster being available for 10min in between, This seems fragile to me. I think it is better to periodically check the status, then if the ping fails, we count 15min since the last successful status check.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:261]: This is not in the scope of this PR. 

`Retry` is the way of keeping a wl for being requeued based on a admission check controller logic.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 9 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go]: This is the best I can do.

Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go]: You mean in this PR / approach on in general we should not track the availability of a worker?

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:261]: I'm wondering if this is needed in this PR. If we moved the logic to workload_controller, it would not be necessary here. 

@alculquicondor which component should be responsible to transitioning Retry to Pending when a workload is evicted: the individual AC controllers, or this is something the workload_controller could handle?

[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:261]: **This is not in the scope of this PR.**

Doing so will prevent other admission check controllers from being able to hold the workload out of the queue.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 9 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go]: In this PR, we should talk about this PR.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:261]: We discussed this offline, and I looked in the KEP, The Retry seems to be associated with `RetryDelayMinutes` which determines the time after which the workload is re-admitted (this could probably be done by workload controller, but TBD). Since we don't support the mechanism yet changing the Retry -> Pending is responsibility of the individual AC controllers. Feel free to resolve.

File: apis/config/v1beta1/configuration_types.go [hunk 18]
```diff
@@ -215,6 +215,11 @@ type MultiKueue struct {
 	// them if their local counterpart no longer exists.
 	// +optional
 	Origin *string `json:"origin,omitempty"`
+
+	// WorkerLostTimeout defines the time a local workload's multikueue admission check state is kept Ready
+	// if the connection with its reserving worker cluster is lost.
```
Comment (alculquicondor) [inline apis/config/v1beta1/configuration_types.go:220]: explain the behavior when the field is left null, or the default.

File: pkg/controller/admissionchecks/multikueue/admissioncheck.go [hunk 19]
```diff
@@ -90,28 +99,25 @@ func (a *ACReconciler) Reconcile(ctx context.Context, req reconcile.Request) (re
 				}
 			}
 		}
+		unusableClustersCount := len(missingClusters) + len(inactiveClusters)
+		if unusableClustersCount > 0 {
+			if unusableClustersCount < len(cfg.Spec.Clusters) {
+				// keep it partially active
+				newCondition.Reason = "PartiallyActive"
```
Comment (alculquicondor) [inline pkg/controller/admissionchecks/multikueue/admissioncheck.go:106]:
```suggestion
				newCondition.Reason = "SomeActiveClusters"
```

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/admissionchecks/multikueue/workload.go:261]: ack

File: apis/config/v1beta1/configuration_types.go (hunk 18 above)
Comment (trasc) [reply apis/config/v1beta1/configuration_types.go:220]: done

File: pkg/controller/admissionchecks/multikueue/admissioncheck.go (hunk 19 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/admissioncheck.go]: done , but why?

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 17 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go]: done

**Comment by mimowo** [comment]
> I'll approve the implementation when @mimowo gives LGTM

Before lgtm for the implementation I would yet like to confirm the implementation approach. IIUC in the current approach the 15min timeout is evaluated between checks, but if there is nothing to trigger the checks in the period of 15min, then it is possible that the cluster is available, but the check after 15min fails (might be even rejected by the P&F). Then, the workload is requeued. I would like to have some mechanism that (1) gives more confidence the cluster is really unavailable, by retrying the check periodically, (2) gives users something more tangible to verify the timeout really passed. I think currently the users don't have a good visibility of when the checks happen, so to verify for them why the requeue happened based on logs will be very hard.

I would like to have some approach that monitors periodically the state of the cluster, say every 30s. The first check unavailable timestamp is recorded in the cluster status. Then, the admission check reconciler verifies if 15min passed since the unavailable timestamp. The timestamp in the status would make it easier to verify / track down the decision for without logs. Also, it would mean that in the 15min period we do more checks, so we don't risk saying "cluster unavailable" if after 15min the call fails for a reason which might be just overloaded API server.

Having said that, I'm ok if we conclude this is overkill, or if we decide this is subject to follow up. However, I would like to have a decision first. WDYT @alculquicondor @mwielgus ?

**Comment by alculquicondor** [comment]
I'm ok with the current behavior for this PR. We can follow up in another one with periodic checks.

**Comment by mimowo** [comment]
Ok, let's go as is, we may iterate on the solution if needed in the future.

/lgtm

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
MuliKueue: Manage worker cluster unavailability
```
